src/Arcus.Messaging.Abstractions/MessageHandling/MessageHandler.cs

```diff
@@ -59,128 +76,117 @@ public static IEnumerable<MessageHandler> SubtractFrom(IServiceProvider serviceP
-                    GetServiceRegistrationTypesFromMicrosoftExtensions(serviceProvider, logger)
-                        .Concat(GetServiceRegistrationTypesFromWebJobs(serviceProvider, logger))
-                        .ToArray();
-
-                if (serviceTypes.Length <= 0)
-                {
-                    logger.LogWarning("No message handlers registrations were found in the service provider");
-                }
-                
-                return serviceTypes;
-            }
-            catch (Exception exception) when (exception is TypeNotFoundException || exception is InvalidCastException)
-            {
-                const string message = "The registered message handlers cannot be retrieved from the service provider because the current version of the dependency injection package doesn't match the expected package used in the Arcus messaging";
-                logger.LogCritical(exception, message);
-                throw new NotSupportedException(message, exception);
-            }
+            Guard.NotNull(messageHandler, nameof(messageHandler), "Requires a message handler implementation to register the handler within the application services");
+
+            ProcessMessageAsync processMessageAsync = DetermineMessageImplementation(messageHandler);
+            logger = logger ?? NullLogger<IMessageHandler<TMessage, TMessageContext>>.Instance;
+            Type messageHandlerType = messageHandler.GetType();
+
+            Func<object, bool> messageFilter = DetermineMessageBodyFilter(messageBodyFilter, messageHandlerType, logger);
+            Func<MessageContext, bool> contextFilter = DetermineMessageContextFilter(messageContextFilter, messageHandlerType, logger);
+
+            return new MessageHandler(
+                messageHandlerInstance: messageHandler,
+                messageHandlerImplementation: processMessageAsync,
+                messageType: typeof(TMessage),
+                messageContextType: typeof(TMessageContext),
+                messageContextFilter: contextFilter,
+                messageBodyFilter: messageFilter,
+                messageBodySerializer: messageBodySerializer,
+                logger: logger);
         }
 
-        private static IEnumerable<Type> GetServiceRegistrationTypesFromMicrosoftExtensions(IServiceProvider serviceProvider, ILogger logger)
+        private static ProcessMessageAsync DetermineMessageImplementation<TMessage, TMessageContext>(IMessageHandler<TMessage, TMessageContext> messageHandler) 
+            where TMessageContext : MessageContext
         {
-            object engine =
-                (serviceProvider.GetOptionalPropertyValue("Engine") 
-                 ?? serviceProvider.GetOptionalFieldValue("_engine"))
-                 ?? GetEngineFromServiceProviderNet6(serviceProvider);
-
-            if (engine is null)
+            return async (rawMessage, generalMessageContext, correlationInfo, cancellationToken) =>
```

Comment:
Yeah, maybe this is too much. It was kinda natural for me to use a functional programming approach here, but I understand that this can seem complex when only thinking in objects. Maybe this can be extracted to a dedicated `DefaultMessageHandler` implementation of some sorts? Then you'll have again your object instead of a function 😉 .

Reply:
A DefaultMessageHandler might make it more readable for me, but it 's OK for me.